Add `StyleEdition` enum and `StyleEditionDefault` trait

ytmimi opened the pull request:
**Note** This does not add the `style_edition` config option to rustfmt. The `StyleEdition` enum will eventually be used to allow users to configure `style_edition`, but for now it's added so we can introduce the the `StyleEditionDefault` trait.

This PR is an alternative to https://github.com/rust-lang/rustfmt/pull/5854 that doesn't use proc macros. 

r? @calebcartwright

calebcartwright commented:
Thanks @ytmimi!
 
I am still planning to try to review this over the weekend. However, also wanted to ensure you'd seen this @fee1-dead given the thoughts you shared on #5854 (though I know you do not have much review capacity currently either)

fee1-dead commented on src/config/style_edition.rs:
```diff
@@ -0,0 +1,69 @@
+#[macro_export]
+macro_rules! style_edition_default {
+    ($ty:ident, $config_ty:ty, $default:expr) => {
+        impl crate::config::style_edition::StyleEditionDefault for $ty {
```
This is very minor but in case we want to have multiple crates changing all the `crate::*` to `$crate:*` would save some time when moving it later.

```suggestion
        impl $crate::config::style_edition::StyleEditionDefault for $ty {
```

fee1-dead commented:
Also please fix the typo when you next push ("triat" -> "trait")

ytmimi commented on src/config/options.rs:
```diff
@@ -468,3 +468,27 @@ pub enum MatchArmLeadingPipe {
+/// [the style guide]: https://doc.rust-lang.org/nightly/style-guide/
+#[config_type]
+pub enum StyleEdition {
+    #[value = "2015"]
```
I don't see why not, but if I'm missing something please let me know. If you happen to find the link for that discussion in cargo I'll be sure to give it a read. I also want to note that we already allow `edition` to be a number.

https://github.com/rust-lang/rustfmt/blob/041f11315923b0e4b0328b3bb6366de9148c7b6b/src/config/options.rs#L417-L436

ytmimi commented on src/config/style_edition.rs:
```diff
@@ -0,0 +1,69 @@
+            }
+        }
+    };
+    ($ty:ident, $config_ty:ty, $default_2015:expr, $default_2024:expr) => {
```
I think this is a good suggestion! I'll update the macro so that it's easier to understand what's going on by looking at the call.

ytmimi commented:
@fee1-dead thanks again for your initial review. I incorporated your suggestions, and fixed the typo that you caught! Please let me know if there's anything else that you'd like me to change and I'll get on that as soon as I can!

fee1-dead reviewed (approved):
LGTM

calebcartwright commented on src/config/style_edition.rs:
```diff
@@ -0,0 +1,69 @@
+            }
+        }
+    };
+    ($ty:ident, $config_ty:ty, Edition2024 => $default_2024:expr, _ => $default_2015:expr) => {
+        impl $crate::config::style_edition::StyleEditionDefault for $ty {
+            type ConfigType = $config_ty;
+
+            fn style_edition_default(
+                style_edition: $crate::config::StyleEdition,
+            ) -> Self::ConfigType {
+                match style_edition {
+                    $crate::config::StyleEdition::Edition2015
+                    | $crate::config::StyleEdition::Edition2018
+                    | $crate::config::StyleEdition::Edition2021 => $default_2015,
```
I wonder if we should make the first three editions roll up to 2021 as the default :thinking: ?

The style edition RFC and subsequent style team policies have established that in cases where an older edition supports syntax that did not exist/did not have rules in the corresponding version of the style guide/edition then the rules from the oldest style edition that _does_ prescribe rules should be used.

Functionally I do not believe it makes a difference here one way or the other, it's almost a code readability thing

ytmimi replied:
> I wonder if we should make the first three editions roll up to 2021 as the default 🤔 ?

Are you suggesting we roll up 2015, 2018, and 2021 enum variants like this:

```rust
enum StyleEdition {
    /// There is no difference between the style in the 2015, 2018, and 2021 editions
    /// so they're rolled up to 2021
    Edition2021,
    Edition2024,
}
```

ytmimi commented:
@calebcartwright thanks for getting this merged!
